### PR TITLE
Diamond

### PR DESCRIPTION
**Pull request by spficklin**
This PR adds the Diamond tool (https://github.com/bbuchfink/diamond) which is an accelerated replacement for blast as a brand new module.   Publishd here:  https://www.nature.com/articles/s41592-021-01101-x  

To generate this module I followed the instructions (can't guarantee I didn't make a mistake), compared it to the existing blast modules for comparison  as it runs very similarly and all tests are working.  

I'm hoping to use this module for a functional annotation workflow I'm helping construct.

## PR checklist

- [X] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [X] Remove all TODO statements.
- [X] Emit the `<SOFTWARE>.version.txt` file.
- [X] Follow the naming conventions.
- [X] Follow the parameters requirements.
- [X] Follow the input/output options guidelines.
- [X] Add a resource `label`
- [X] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [X] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [X] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [X] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: modules/diamond/blastp/main.nf [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        container "quay.io/biocontainers/diamond:2.0.9--hdcc8f71_0"
+    }
+
+    input:
+    tuple val(meta), path(fasta)
+    path  db
+
+    output:
+    tuple val(meta), path('*.diamond_blastp.txt'), emit: txt
+    path '*.version.txt', emit: version
```
Comment (drpatelh) [inline modules/diamond/blastp/main.nf:29]:
```suggestion
    tuple val(meta), path('*.txt'), emit: txt
    path '*.version.txt'          , emit: version
```

File: modules/diamond/blastp/main.nf [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+
```
Comment (drpatelh) [inline modules/diamond/blastp/main.nf:34]:
```suggestion
```

File: modules/diamond/blastp/main.nf [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    """
+    DB=`find -L ./ -name "*.dmnd" | sed 's/.dmnd//'`
+    diamond blastp \\
+        --threads $task.cpus \\
```
Comment (drpatelh) [inline modules/diamond/blastp/main.nf:38]:
```suggestion

    diamond \\
        blastp \\
        --threads $task.cpus \\
```

File: modules/diamond/blastp/main.nf [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        --db \$DB \\
+        --query $fasta \\
+        $options.args \\
+        --out ${prefix}.diamond_blastp.txt
+    echo \$(diamond --version 2>&1) | tail -n 1 | sed 's/^diamond version //' > ${software}.version.txt
```
Comment (drpatelh) [inline modules/diamond/blastp/main.nf:43]:
```suggestion
        --out ${prefix}.txt

    echo \$(diamond --version 2>&1) | tail -n 1 | sed 's/^diamond version //' > ${software}.version.txt
```

File: modules/diamond/blastp/meta.yml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        description: Input fasta file containing query sequences
+        pattern: "*.{fa,fasta}"
+  - db:
+        type: directory
+        description: Directory containing the protein blast database
+        pattern: "*"
+
+output:
+  - txt:
+    type: file
```
Comment (drpatelh) [inline modules/diamond/blastp/meta.yml:34]: Indents are off in this section.

File: modules/diamond/blastx/main.nf [hunk 6]
```diff
@@ -0,0 +1,44 @@
+// Import generic module functions
+include { initOptions; saveFiles; getSoftwareName } from './functions'
+
+params.options = [:]
+options    = initOptions(params.options)
```
Comment (drpatelh) [inline modules/diamond/blastx/main.nf:5]:
```suggestion
params.options = [:]
options        = initOptions(params.options)
```

File: modules/diamond/blastx/main.nf [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    """
+    DB=`find -L ./ -name "*.dmnd" | sed 's/.dmnd//'`
+    diamond blastx \\
+        --threads $task.cpus \\
+        --db \$DB \\
+        --query $fasta \\
+        $options.args \\
+        --out ${prefix}.diamond_blastx.txt
```
Comment (drpatelh) [inline modules/diamond/blastx/main.nf:41]:
```suggestion
        --out ${prefix}.txt

```

File: modules/diamond/makedb/main.nf [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    path fasta
+
+    output:
+    path "${fasta}.dmnd", emit: db
+    path '*.version.txt', emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    """
+    diamond makedb \\
```
Comment (drpatelh) [inline modules/diamond/makedb/main.nf:33]:
```suggestion
    diamond \\
        makedb \\
```

File: modules/diamond/makedb/main.nf [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        --threads $task.cpus \\
+        --in  $fasta \\
+        -d $fasta \\
+        $options.args
+    echo \$(diamond --version 2>&1) | tail -n 1 | sed 's/^diamond version //' > ${software}.version.txt
```
Comment (drpatelh) [inline modules/diamond/makedb/main.nf:38]:
```suggestion

    echo \$(diamond --version 2>&1) | tail -n 1 | sed 's/^diamond version //' > ${software}.version.txt
```

File: tests/modules/diamond/blastp/main.nf [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+include { DIAMOND_MAKEDB } from '../../../../modules/diamond/makedb/main.nf' addParams( options: [:] )
+include { DIAMOND_BLASTP } from '../../../../modules/diamond/blastp/main.nf' addParams( options: [:] )
+
+workflow test_diamond_blastp {
+
+    db = [ file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true) ]
+    fasta = [ file(params.test_data['sarscov2']['genome']['transcriptome_fasta'], checkIfExists: true) ]
+
+    DIAMOND_MAKEDB ( db )
+    DIAMOND_BLASTP( [ [id:'test'], fasta ], DIAMOND_MAKEDB.out.db )
```
Comment (drpatelh) [inline tests/modules/diamond/blastp/main.nf:14]:
```suggestion
    DIAMOND_BLASTP ( [ [id:'test'], fasta ], DIAMOND_MAKEDB.out.db )
```

File: tests/modules/diamond/blastx/main.nf [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+include { DIAMOND_MAKEDB } from '../../../../modules/diamond/makedb/main.nf' addParams( options: [:] )
+include { DIAMOND_BLASTX } from '../../../../modules/diamond/blastx/main.nf' addParams( options: [:] )
+
+workflow test_diamond_blastx {
+
+    db = [ file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true) ]
+    fasta = [ file(params.test_data['sarscov2']['genome']['transcriptome_fasta'], checkIfExists: true) ]
+
+    DIAMOND_MAKEDB ( db )
+    DIAMOND_BLASTX( [ [id:'test'], fasta ], DIAMOND_MAKEDB.out.db )
```
Comment (drpatelh) [inline tests/modules/diamond/blastx/main.nf:14]:
```suggestion
    DIAMOND_BLASTX ( [ [id:'test'], fasta ], DIAMOND_MAKEDB.out.db )
```

File: tests/modules/diamond/blastx/main.nf [hunk 12]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { DIAMOND_MAKEDB } from '../../../../modules/diamond/makedb/main.nf' addParams( options: [:] )
+include { DIAMOND_BLASTX } from '../../../../modules/diamond/blastx/main.nf' addParams( options: [:] )
```
Comment (drpatelh) [inline tests/modules/diamond/blastx/main.nf:6]:
```suggestion
include { DIAMOND_MAKEDB } from '../../../../modules/diamond/makedb/main.nf' addParams( options: [:]                           )
include { DIAMOND_BLASTX } from '../../../../modules/diamond/blastx/main.nf' addParams( options: [ suffix: '.diamond_blastx' ] )
```

File: tests/modules/diamond/blastp/main.nf [hunk 13]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { DIAMOND_MAKEDB } from '../../../../modules/diamond/makedb/main.nf' addParams( options: [:] )
+include { DIAMOND_BLASTP } from '../../../../modules/diamond/blastp/main.nf' addParams( options: [:] )
```
Comment (drpatelh) [inline tests/modules/diamond/blastp/main.nf:6]:
```suggestion
include { DIAMOND_MAKEDB } from '../../../../modules/diamond/makedb/main.nf' addParams( options: [:]                           )
include { DIAMOND_BLASTP } from '../../../../modules/diamond/blastp/main.nf' addParams( options: [ suffix: '.diamond_blastp' ] )
```

**Review by drpatelh — CHANGES_REQUESTED** [review]
Hi @spficklin ! Thanks for adding these modules! Was going to ask if you could split this PR into one module per PR but I reviewed here anyway so maybe one for next time :) Looks great overall just a few minor changes. I don't have permissions to push to your org otherwise I would have tried to get it in.

File: modules/diamond/blastp/meta.yml [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    description: File containing blastp hits
+    pattern: "*.{blastp.txt}"
+  - version:
+    type: file
+    description: File containing software version
+    pattern: "*.{version.txt}"
```
Comment (drpatelh) [inline modules/diamond/blastp/meta.yml:40]:
```suggestion
        pattern: "*"

output:
  - txt:
        type: file
        description: File containing blastp hits
        pattern: "*.{blastp.txt}"
  - version:
        type: file
        description: File containing software version
        pattern: "*.{version.txt}"
```

**Comment by spficklin** [comment]
Thanks @drpatelh for the suggested changes!  I've merged them into the PR.  I will be aware next time separate out the modules into separate PRs even if they are for the same tool.  

I think I've merged in all your suggested changes I'm happy to make more changes if there's anything else or I missed something.

**Review by drpatelh — APPROVED** [review]

**Comment by drpatelh** [comment]
No this is 👌🏽 Thank you :)

**Comment by spficklin** [comment]
This was my first module submission and the process was great! Thank you.

**Pull request merged**